### PR TITLE
Handle links in Custom UI and Notifications

### PR DESCRIPTION
**Pull request by GuillaumeRx**
This adds the ability to display links in dialogs, notifications and tx-insights.

- Add links verification for format and against Phishing List
- Add upper link verification in notify and dialogs
- Assert snap response type for `onTransaction` handler
- Update Simulator

Fixes: MetaMask/MetaMask-planning/issues/1200 MetaMask/MetaMask-planning/issues/551 MetaMask/MetaMask-planning/issues/553 #1792

### CODE REVIEW

**Comment by GuillaumeRx** [comment]
@metamaskbot update-pr

**Comment by GuillaumeRx** [comment]
@socket-security ignore @metamask/phishing-controller

owned dependency

File: packages/snaps-controllers/package.json [hunk 1]
```diff
@@ -134,5 +136,17 @@
...
+    "allowScripts": {
+      "@metamask/phishing-controller>@metamask/controller-utils>ethereumjs-util>ethereum-cryptography>keccak": false,
+      "@metamask/phishing-controller>@metamask/controller-utils>ethereumjs-util>ethereum-cryptography>secp256k1": false,
+      "@swc/core": false,
+      "@wdio/browser-runner>@originjs/vite-plugin-commonjs>esbuild": false,
+      "@wdio/cli>@wdio/utils>edgedriver": false,
+      "esbuild": false,
+      "wdio-chromedriver-service>chromedriver": false,
+      "wdio-geckodriver-service>geckodriver": false
+    }
```
Comment (Mrtenz) [inline packages/snaps-controllers/package.json:150]: This shouldn't be in the `snaps-controllers` `package.json`.

File: packages/snaps-rpc-methods/src/restricted/invokeSnap.ts [hunk 2]
```diff
@@ -14,7 +14,7 @@ import type {
   RequestedSnapPermissions,
   InstallSnapsResult,
 } from '@metamask/snaps-utils';
-import { HandlerType, SnapCaveatType } from '@metamask/snaps-utils';
+import { SnapCaveatType, HandlerType } from '@metamask/snaps-utils';
```
Comment (Mrtenz) [inline packages/snaps-rpc-methods/src/restricted/invokeSnap.ts:17]: ?

File: packages/snaps-rpc-methods/src/utils.ts [hunk 3]
```diff
@@ -4,7 +4,7 @@ import type {
   SLIP10PathNode,
 } from '@metamask/key-tree';
 import { SLIP10Node } from '@metamask/key-tree';
-import type { MagicValue } from '@metamask/snaps-utils';
+import { type MagicValue } from '@metamask/snaps-utils';
```
Comment (Mrtenz) [inline packages/snaps-rpc-methods/src/utils.ts:7]: ?

File: packages/snaps-simulator/src/features/renderer/components/Text.tsx [hunk 4]
```diff
@@ -15,13 +15,26 @@ export const Text: FunctionComponent<TextProps> = ({ node, id }) => {
...
     <ReactMarkdown
-      allowedElements={['p', 'strong', 'em']}
+      allowedElements={['p', 'strong', 'em', 'a']}
       components={{
         p: ({ children: value }) => (
           <ChakraText fontFamily="custom" fontSize="sm" paddingBottom="1">
             {value}
           </ChakraText>
         ),
+        // TODO: Update styling
```
Comment (Mrtenz) [inline packages/snaps-simulator/src/features/renderer/components/Text.tsx:25]: Are you planning on doing that in this PR, or separate? If separate, please create an issue for it.

File: packages/snaps-ui/src/validation.ts [hunk 5]
```diff
@@ -25,3 +25,72 @@ export function isComponent(value: unknown): value is Component {
...
 }
+
+const LINK_REGEX = /(?<protocol>[a-z]+:\/?\/?)(?<host>\S+?(?:\.[a-z]+)+)/giu;
+
+enum AlloweProtocols {
+  Https = 'https:',
+  Mailto = 'mailto:',
+}
+
+const ALLOWED_PROTOCOLS = Object.values(AlloweProtocols);
```
Comment (Mrtenz) [inline packages/snaps-ui/src/validation.ts:36]: Seems easier to just use an array here. That also removes the need for the type cast further on.
```suggestion
const ALLOWED_PROTOCOLS = [
  'https:',
  'mailto:',
];
```

File: packages/snaps-ui/src/validation.ts [hunk 6]
```diff
@@ -25,3 +25,72 @@ export function isComponent(value: unknown): value is Component {
...
+          ALLOWED_PROTOCOLS.includes(url.protocol as AlloweProtocols),
+          `protocol must be one of: ${ALLOWED_PROTOCOLS.join(', ')}`,
+        );
+
+        const hostname =
+          url.protocol === AlloweProtocols.Mailto
+            ? url.pathname.split('@')[1]
+            : url.hostname;
+
+        assert(!isOnPhishingList(hostname), 'detected as phishing');
```
Comment (Mrtenz) [inline packages/snaps-ui/src/validation.ts:63]:
```suggestion
        assert(!isOnPhishingList(hostname), 'The specified URL is not allowed.');
```

File: packages/snaps-ui/src/validation.ts [hunk 7]
```diff
@@ -25,3 +25,72 @@ export function isComponent(value: unknown): value is Component {
...
+      } catch (error) {
+        throw new Error(
+          `Invalid URL: ${
+            error instanceof AssertionError ? error.message : 'invalid sintax'
```
Comment (Mrtenz) [inline packages/snaps-ui/src/validation.ts:67]:
```suggestion
            error instanceof AssertionError ? error.message : 'Invalid syntax.'
```

When is this thrown?

File: packages/snaps-ui/src/validation.ts [hunk 8]
```diff
@@ -25,3 +25,72 @@ export function isComponent(value: unknown): value is Component {
...
+  const { type } = component;
+  if (type === NodeType.Panel) {
+    component.children.forEach((node) =>
+      assertUILinksAreSafe(node, isOnPhishingList),
+    );
+  }
+
+  if (component.type === NodeType.Text) {
+    assertLinksAreSafe(component.value, isOnPhishingList);
+  }
```
Comment (Mrtenz) [inline packages/snaps-ui/src/validation.ts:95]: How about other components with values, such as `Copyable`? Maybe we should just check any component with a `value` (and maybe the other fields too).

File: packages/snaps-utils/src/handlers.ts [hunk 9]
```diff
@@ -113,17 +88,19 @@ export enum SeverityLevel {
   Critical = 'critical',
 }
 
+export const OnTransactionResponseStruct = object({
+  content: ComponentStruct,
+  severity: optional(union([literal(SeverityLevel.Critical)])),
```
Comment (Mrtenz) [inline packages/snaps-utils/src/handlers.ts:93]:
```suggestion
  severity: optional(literal(SeverityLevel.Critical)),
```

File: packages/snaps-simulator/src/features/renderer/components/Text.tsx (hunk 4 above)
Comment (GuillaumeRx) [reply packages/snaps-simulator/src/features/renderer/components/Text.tsx:25]: Yup, I still need to see the design with @eriknson

File: packages/snaps-ui/src/validation.ts (hunk 7 above)
Comment (GuillaumeRx) [reply packages/snaps-ui/src/validation.ts:67]: `Invalid syntax` is thrown if there's an error with `new URL()`

File: packages/snaps-simulator/src/features/renderer/components/Text.tsx (hunk 4 above)
Comment (Mrtenz) [reply packages/snaps-simulator/src/features/renderer/components/Text.tsx:25]: Ok, can you link the issue here and in the code?

File: packages/snaps-ui/src/validation.ts (hunk 7 above)
Comment (Mrtenz) [reply packages/snaps-ui/src/validation.ts]: Maybe "Unable to parse URL."?

File: packages/snaps-ui/src/validation.ts [hunk 10]
```diff
@@ -25,3 +30,70 @@ export function isComponent(value: unknown): value is Component {
...
+ *
+ * @param component - The custom UI component.
+ * @param isOnPhishingList - The function that checks the link against the phishing list.
+ */
+export function assertUILinksAreSafe(
+  component: Component,
+  isOnPhishingList: (url: string) => boolean,
+) {
+  const { type } = component;
+  if (type === NodeType.Panel) {
```
Comment (FrederikBolding) [inline packages/snaps-ui/src/validation.ts:90]: @Mrtenz Is there an easier way we could do this with superstruct? If we could avoid manually traversing the components I would say that is ideal.

File: packages/snaps-simulator/src/features/renderer/components/Text.tsx (hunk 4 above)
Comment (GuillaumeRx) [reply packages/snaps-simulator/src/features/renderer/components/Text.tsx:25]: https://github.com/MetaMask/snaps/issues/1881

File: packages/snaps-ui/src/validation.ts (hunk 10 above)
Comment (GuillaumeRx) [reply packages/snaps-ui/src/validation.ts:92]: I thought of creating a new refinement util for superstruct but it seems to be very painfull to always provide the function to the controller call everytime we need to validate the struct

Comment (Mrtenz) [reply packages/snaps-ui/src/validation.ts:92]: Yeah that would be ideal, but I don't think there is a good way to provide the validation function to the structs.

**Review by Mrtenz — APPROVED** [review]

**Review by Mrtenz — APPROVED** [review]

**Pull request merged**